src/mock/quick_p2p/tests.rs

```diff
@@ -527,6 +523,22 @@ impl Agent {
     fn expect_none(&self) {
         assert_eq!(self.rx.try_recv(), Err(TryRecvError::Empty));
     }
+
+    fn received_messages(&self, src_addr: &SocketAddr) -> BTreeSet<Bytes> {
+        let mut received_messages = BTreeSet::new();
+        while let Ok(Event::NewMessage { peer_addr, msg }) = self.rx.try_recv() {
+            assert_eq!(peer_addr, *src_addr);
+            let _ = received_messages.insert(msg);
+        }
+        received_messages
+    }
+}
+
+fn expected_messags_received(expected: Vec<Bytes>, mut received: BTreeSet<Bytes>) {
```

Comment:
Why take 2 different types for things that are conceptually similar?
This function should probably take two arguments of the same type, `assert!` they're equal and reset received to `ThatCollection::new()`.

In particular, I see a `to_vec()` at line 314 which makes me think the `Vec` isn't really a natural representation for this `expected` argument.

Reply:
first, we don't expect these two are equal, as the received order could be different.
second, I made these two different purposely to diff `expected` and `received`

Reply:
uhm. If you want for the types to be different for type-safety, I would suggest to be explicit and create a `struct Expected (BTreeSet<Bytes>)` and a `struct Actual(BtreeSet<Bytes>)`. Probably overkill, but better than changing one to a Vec somewhat arbitrarily imo.

Reply:
Adam had the same concern and we had a discussion at https://github.com/maidsafe/routing/pull/1688#discussion_r291955790
And reached an agreement.

Reply:
Uhm. I'm not sure it's exactly the same concern. To be clear, I think the implementation now is better than what was there before. I am simply still unsure about passing a `Vec` for the `sent` instead of a `BTreeSet`. If I'm not mistaken, we're converting from `BTreeSet` to `vec` and then vice-versa.

Reply:
> we're converting from BTreeSet to vec and then vice-versa

No we don't. The vection is constructed directly in the testing code:
```
    let msgs = [
        Bytes::from_static(b"message 0 from A"),
        Bytes::from_static(b"message 1 from A"),
        Bytes::from_static(b"message 2 from A"),
    ];
```

Reply:
Uhm. OK. I reread the thread. I had missed something from the msg before last. I may have been confused about the type at the call site. I'm OK as is.